hashin.py

```diff
@@ -196,6 +158,30 @@ def run_packages(
         # We do that by modifying only the `name` part of the `Requirement` instance.
         req.name = package
 
+        new_version_specifier = SpecifierSet("=={}".format(data["version"]))
+
+        if previous_versions and previous_versions.get(str(req).lower()):
+            # We have some form of previous version and a new version.
+            # If they' already equal, just skip this one.
+            if previous_versions[str(req).lower()] == new_version_specifier:
+                continue
+
+        if interactive:
+            try:
+                if not interactive_upgrade_request(
+                    package,
+                    previous_versions[str(req).lower()],
+                    new_version_specifier,
+                    print_header=first_interactive,
+                ):
+                    first_interactive = False
+                    continue
+                first_interactive = False
+            except InteractiveAll:
```

Comment:
My general rule is to only use exceptions for exceptional cases. "All" and "Quit" don't seem exceptional to me. I think I would make `interactive_upgrade_request` return a set of constants for all the cases, instead of returning a boolean in some cases and raising exception in others.

Reply:
Yeah, I tend to agree. ...in hindsight. I think I fell into the trap of doing it like this because I had originally not planned to use "all" and "quit" and when I added those it didn't match the way I called the `interactive_upgrade_request` function.